source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py

```diff
@@ -169,35 +169,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
 		# However, in build 17666 and later, child count is the same for both emoji panel and hardware keyboard candidates list.
 		# Thankfully first child automation ID's are different for each modern input technology.
 		# However this event is raised when the input panel closes.
-		if obj.firstChild is None:
+		if firstChild is None:
 			return
 		# #9104: different aspects of modern input panel are represented by automation iD's.
-		childAutomationID = obj.firstChild.UIAElement.cachedAutomationID
+		childAutomationId = firstChild.UIAAutomationId
 		# Emoji panel for build 16299 and 17134.
 		# This event is properly raised in build 17134.
-		if winVersion.winVersion.build <= 17134 and childAutomationID in ("TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl", "TEMPLATE_PART_ExpressiveInputFullViewFuntionBarCloseButton"):
-			self.event_UIA_elementSelected(obj.lastChild.firstChild, nextHandler)
+		if not winVersion.isWin10(version=1809) and childAutomationId in self._classicEmojiPanelAutomationIds:
+			eventHandler.executeEvent("UIA_elementSelected", obj.lastChild.firstChild)
 		# Handle hardware keyboard suggestions.
 		# Treat it the same as CJK composition list - don't announce this if candidate announcement setting is off.
-		elif childAutomationID == "CandidateWindowControl" and config.conf["inputComposition"]["autoReportAllCandidates"]:
+		elif (
+			childAutomationId == "CandidateWindowControl"
+			and config.conf["inputComposition"]["autoReportAllCandidates"]
+		):
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.firstChild.firstChild, nextHandler)
+				eventHandler.executeEvent("UIA_elementSelected", firstChild.firstChild.firstChild)
 			except AttributeError:
 				# Because this is dictation window.
 				pass
-		# Emoji panel in build 17666 and later (unless this changes).
-		elif childAutomationID == "TEMPLATE_PART_ExpressionGroupedFullView":
+		# Emoji panel in Version 1809 (specifically, build 17666) and later.
+		elif childAutomationId == "TEMPLATE_PART_ExpressionGroupedFullView":
 			self._emojiPanelJustOpened = True
+			# #10377: on some systems, there is something else besides grouping controls,
+			# so another child control must be used.
+			emojisList = firstChild.children[-2]
+			if emojisList.UIAAutomationId != "TEMPLATE_PART_Items_GridView":
+				emojisList = emojisList.previous
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+				# Avoid announcing skin tone modifiers if possible.
+				emojiItem = emojisList.firstChild.firstChild

```

Comment:
Again, are we sure that there is a first child that has a first child? I'm afraid this can lead to breakage when this changes in future.

Reply:
Yes, for now emoji panel consists of the emoji list, a grouping (first child), and the actual emojis list unless people emoji is selected, in which case the third level will be skin tone modifiers list object. We need to come back to this if we get a confirmation that the new modern keyboard UI significantly changes this. The upcoming commit will check if the first child object is present for each level as suggested.